demo/apisix/tmail-apisix-plugin-runner/src/main/java/com/linagora/apisix/plugin/RewriteXUserFilter.java

```diff
@@ -0,0 +1,72 @@
+package com.linagora.apisix.plugin;
+
+import static com.linagora.apisix.plugin.TokenRevokedFilter.makeUnAuthorizedRequest;
+
+import java.io.IOException;
+import java.util.Base64;
+import java.util.Locale;
+import java.util.Optional;
+
+import org.apache.apisix.plugin.runner.HttpRequest;
+import org.apache.apisix.plugin.runner.HttpResponse;
+import org.apache.apisix.plugin.runner.filter.PluginFilter;
+import org.apache.apisix.plugin.runner.filter.PluginFilterChain;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+@Component
+public class RewriteXUserFilter implements PluginFilter {
+    public static final String TMAIL_X_USER_HEADER_NAME = "X-User";
+    private static final String PRE_FILTER = "pre";
+    private static final String POST_FILTER = "post";
+    private final Logger logger = LoggerFactory.getLogger("RewriteXUserPlugin");
+    private final ObjectMapper json;
+
+    public RewriteXUserFilter() {
+        this.json = new ObjectMapper();
+    }
+
+    @Override
+    public String name() {
+        return "RewriteXUserFilter";
+    }
+
+    @Override
+    public void filter(HttpRequest request, HttpResponse response, PluginFilterChain chain) {
+        final String mode = request.getConfig(this);
+        logger.debug("Received a new request. Filter mode: {}", mode);
+        if (mode.equalsIgnoreCase(PRE_FILTER)) {
+            xUserHeaderConditionCheck(request, response);
+        } else if (mode.equalsIgnoreCase(POST_FILTER)) {
+            Optional.ofNullable(request.getHeader("X-Userinfo"))
+                .or(() -> Optional.ofNullable(request.getHeader("x-userinfo")))
+                .flatMap(this::extractUserFromUserInfo)
+                .ifPresent(user -> {
+                    request.setHeader(TMAIL_X_USER_HEADER_NAME, user);
+                    logger.debug("Rewrite X-User header. {}", user);
+                });
+        }
+        chain.filter(request, response);
+    }
+
+    private void xUserHeaderConditionCheck(HttpRequest request, HttpResponse response) {
+        Optional.ofNullable(request.getHeader(TMAIL_X_USER_HEADER_NAME))
+            .or(() -> Optional.ofNullable(request.getHeader(TMAIL_X_USER_HEADER_NAME.toLowerCase(Locale.US))))
+            .ifPresent(deniedHeader -> {
+                makeUnAuthorizedRequest(request, response);
+                logger.debug("The client request has X-User header has been denied. {}", deniedHeader);
+            });
+    }
+
+    public Optional<String> extractUserFromUserInfo(String userInfo) {
+        try {
+            return Optional.of(json.readTree(Base64.getDecoder().decode(userInfo)).get("sub").asText());
```

Comment:
Can we make the extracted field be configurable BTW?

Reply:
Yes, we can, but now make it eazy, 
`sub` is a core field of JWT

Reply:
This really makes sense. 
userInfo of Keycloak and LemonLdap is difference format